pyrmq/consumer.py

```diff
@@ -100,13 +129,73 @@ def __send_reconnection_error_message(self, retry_count, error) -> None:
 
         logger.exception(error)
 
+    def __send_consume_error_message(self, retry_count: int, error: Exception) -> None:
+        """
+        Send error message to your preferred location.
+        :param retry_count: Amount retries the Consumer tried before sending an error message.
+        :param error: Error that prevented the Consumer from processing the callback.
+        """
+        message = (
+            f"Service tried to consume message **{retry_count}** times "
+            f"but still failed."
+            f"\n{repr(error)}"
+        )
+        if self.error_callback:
+            self.error_callback(message)
+
+        logger.exception(error)
+
     def __create_connection(self) -> BlockingConnection:
         """
         Creates a pika BlockingConnection from the given connection parameters.
         """
         return BlockingConnection(self.connection_parameters)
 
-    def _consume_message(self, channel, method, properties, data) -> None:
+    def _compute_expiration(self, retry_count: int) -> int:
+        """
+        Computes message expiration time from the retry queue in seconds.
+        """
+        b = self.retry_backoff_base
+        n = self.retry_delay * 1000
+        return b ** (retry_count - 1) * n  # 5, 10, 20, 40, 80
+
+    def _publish_to_retry_queue(
+        self, data: dict, properties, retry_reason: Exception = None
+    ) -> None:
+        """
+        Publishes message to retry queue with the appropriate metadata in the headers.
+        """
+        headers = properties.headers or {}
+        attempt = (headers.get("x-attempt", 0) + 1) or 1
+        self.__send_consume_error_message(attempt, retry_reason)
+
+        if attempt > self.max_retries:
+            return
+
+        expiration = self._compute_expiration(attempt)
+        now = datetime.now()
+        next_attempt = now + timedelta(seconds=(expiration // 1000))
+        message_properties = {
+            "expiration": str(expiration),
+            "headers": {
+                **headers,
+                "x-attempt": attempt,
+                "x-max-attempts": self.max_retries,
+                "x-created-at": headers.get("x-created-at", now.isoformat()),
+                "x-retry-reason": str(retry_reason),
+                "x-next-attempt": next_attempt.isoformat(),
+            },
+        }
+        for i in range(1, attempt + 1):
+            message_properties["headers"][f"x-attempt-{i}"] = message_properties[
+                "headers"
+            ].get(f"x-attempt-{i}", now.isoformat())
+
+        self.publisher.publish(
+            data, message_properties={**properties.__dict__, **message_properties}
```

Comment:
How about including `properties.__dict__` already in `message_properties` so we don't have to spread the two here?